Clear SAML IdP session on logout

flyinghermit opened the pull request:
Clear SAML IdP session cookie and Delete all active SAML user sessions on logout.

Fixes https://github.com/gravitational/teleport/issues/35915

flyinghermit committed:
clear saml session on logout

flyinghermit committed:
saml idp session logout test

Tener commented on lib/web/sessions.go:
```diff
@@ -862,6 +862,9 @@ func (s *sessionCache) invalidateSession(ctx context.Context, sctx *SessionConte
 	if err := clt.DeleteUserAppSessions(ctx, &proto.DeleteUserAppSessionsRequest{Username: sctx.GetUser()}); err != nil {
 		return trace.Wrap(err)
 	}
+	if err := clt.DeleteUserSAMLIdPSessions(ctx, sctx.GetUser()); err != nil {
+		return trace.Wrap(err)
+	}
```
Why do we need to delete all sessions? I'd expect both cookies and sessions to be on per-browser basis, yet here we delete everything, which means logout in one browser will invalidate all other active sessions. This may be a nice feature to have, but I think we should be more selective by default.

flyinghermit replied:
Hmmm I think you are right. Since this is user logging out from IdP, I thought it would be good to remove all active sessions. But for such feature, looks like it would be better to provide "clear session from all browser" kind of option.

flyinghermit replied:
@Tener here's the commit which now clears user's current session instead of all SAML session. https://github.com/gravitational/teleport/pull/37619/commits/dd5cd08f5af993fa2c7b4eded05b9de7c9d4a481

I've used SAML session cookie to get the session ID as currently, the `websession` state is not updated with SAML session after SAML auth. I assessed to have the `websession` state updated but looks like that demands some more work outside the scope of this PR.  And I believe, retrieving session ID from cookie value for logout doesn't seem fallible option either. Let me know if you think other way!

codingllama commented on lib/web/session/cookie.go:
```diff
@@ -84,6 +84,10 @@ func ClearCookie(w http.ResponseWriter) {
+
+	// SAMLSessionCookieName is the name of the SAML IdP session cookie.
+	// This cookie is set by SAML IdP after a successful SAML authentication.
+	SAMLSessionCookieName = "__Host-saml_session"
```
Similarly as above, let's avoid mixing cookies with different purposes in this file.

codingllama commented on lib/web/apiserver_test.go:
```diff
@@ -8965,6 +8965,59 @@ func TestLogout(t *testing.T) {
+	env := newWebPack(t, 2)
+
+	// create a logged in user for proxy 1.
+	user := "llama"
```
nit:

```suggestion
	const user = "llama"
```

codingllama commented on lib/web/apiserver_test.go:
```diff
@@ -8965,6 +8965,59 @@ func TestLogout(t *testing.T) {
+	user := "llama"
+	pack := env.proxies[0].authPack(t, user, nil /* roles */)
+
+	samlSessionID := "saml_session_id"
```
```suggestion
	const samlSessionID = "saml_session_id"
```

codingllama commented on lib/web/apiserver_test.go:
```diff
@@ -8965,6 +8965,59 @@ func TestLogout(t *testing.T) {
+	require.ErrorContains(t, err, `key "/saml_idp/sessions/saml_session_id" is not found`)
+}
+
+func matchSAMLSessionCookie(cookies []*http.Cookie) bool {
```
Suggestion:

```suggestion
func hasEmptySAMLSessionCookie(cookies []*http.Cookie) bool {
```

codingllama commented on lib/web/apiserver.go:
```diff
@@ -2168,7 +2168,10 @@ func (h *Handler) logout(ctx context.Context, w http.ResponseWriter, sctx *Sessi
 		return trace.Wrap(err)
 	}
-	websession.ClearCookie(w)
+	// Clear Web UI session cookie
+	websession.ClearCookie(w, websession.CookieName)
+	// Clear SAML IdP session cookie
+	websession.ClearCookie(w, websession.SAMLSessionCookieName)
```
Do we always expect these to be cleared together? If yes, then maybe pull both clears into a single "clearSessionCookies" func and reuse it?

flyinghermit replied:
https://github.com/gravitational/teleport/pull/37619/commits/96be22db00fce017e018a5f4f98176e2310c7627

flyinghermit commented on lib/web/session/cookie.go:
```diff
@@ -73,9 +73,9 @@ func SetCookie(w http.ResponseWriter, user, sid string) error {
 
 // ClearCookie wipes the session cookie to invalidate the user session.
-func ClearCookie(w http.ResponseWriter) {
+func ClearCookie(w http.ResponseWriter, cookieName string) {
```
Sure I can add another `clearSAMLCookie` func here but I don't think this warrant an entirely new package.

flyinghermit replied:
I moved SAML coookie code to seperate package `lib/idp/saml` (the package structure is same as we have in the `e` repo) https://github.com/gravitational/teleport/pull/37619/commits/96be22db00fce017e018a5f4f98176e2310c7627

codingllama replied:
Thanks, looks good!

flyinghermit committed:
rename test func and use const for declaration

flyinghermit committed:
clear current browser session on logout

flyinghermit committed:
- add lib/idp/saml package to contain saml idp session cookie functions
- revert web/session package to remove SAML session cookie
- add clearSessionCookie func to clear web and saml session

flyinghermit committed:
fix linter

codingllama commented on lib/idp/saml/session.go:
```diff
@@ -0,0 +1,52 @@
+)
+
+// SetCookie set's SAML session cookie named __Host-saml_session.
+func SetCookie(w http.ResponseWriter, sessionID string, maxAge int) {
```
```suggestion
func SetCookie(w http.ResponseWriter, sessionID string, maxAge time.Duration) {
```

This way there's no question about what the time unit is.

mdwn replied:
This isn't as intuitive as it seems because `maxAge` in the http Cookie is annoying fixed to seconds. No reason we can't use duration, but that's the reason for this choice here.

codingllama replied:
Let's do `maxAgeSeconds int`, then, so it's clear what the unit is.

codingllama commented on lib/web/apiserver.go:
```diff
@@ -2148,15 +2149,26 @@ func clientMetaFromReq(r *http.Request) *auth.ForwardedClientMetadata {
+	// with SAML details.
+	samlSessionCookie, err := r.Cookie(samlidp.SAMLSessionCookieName)
+	if err != nil {
+		h.log.Debug("SAML IdP session not found for deletion for user: %s", ctx.GetUser())
```
How unexpected is this condition? A comment clarifying would help.

flyinghermit replied:
My initial plan was to swallow the error completely. For generic Teleport installation, this is just going to be log noise, I've removed this in https://github.com/gravitational/teleport/pull/37619/commits/7f451952eb724dda594c4e7e45eb8746b0d35e13

codingllama commented on lib/web/apiserver.go:
```diff
@@ -2148,15 +2149,26 @@ func clientMetaFromReq(r *http.Request) *auth.ForwardedClientMetadata {
+	// with SAML details.
+	samlSessionCookie, err := r.Cookie(samlidp.SAMLSessionCookieName)
+	if err != nil {
+		h.log.Debug("SAML IdP session not found for deletion for user: %s", ctx.GetUser())
```
Use structured logging (also a "Debug" statement with format directive is an error).

```suggestion
		h.log.
			WithError(err).
			WithField("user", ctx.GetUser()).
			Debug("SAML IdP session not found for deletion for user")
```

codingllama commented on lib/web/apiserver.go:
```diff
@@ -2148,15 +2149,26 @@ func clientMetaFromReq(r *http.Request) *auth.ForwardedClientMetadata {
+	if err != nil {
+		h.log.Debug("SAML IdP session not found for deletion for user: %s", ctx.GetUser())
+	} else if samlSessionCookie != nil && samlSessionCookie.Value != "" {
+		if ctx.cfg.Session.GetSAMLSession() == nil {
```
Why does this need to happen here?

Would it be possible for a SAML session to already exist, but point at a different cookie?

flyinghermit replied:
> Why does this need to happen here?

Only because this value is never set until now. This PR sets it during logout, but that too happens only if SAML session cookie is present.

> Would it be possible for a SAML session to already exist, but point at a different cookie?

In the same browser? No. Not unless user tries to spoof session cookie value.
If that happens then the SAML IdP detects session ID mismatch and fails the login process.

codingllama replied:
Sounds good. Could you add a comment with some of that background?

codingllama replied:
Bump for this one? I'll leave it approve so I don't block you further.

flyinghermit replied:
https://github.com/gravitational/teleport/pull/37619/commits/dc53d1730a81b43b7a7cab402e5a1c4fc7678aab

codingllama commented on lib/web/apiserver.go:
```diff
@@ -2168,11 +2180,18 @@ func (h *Handler) logout(ctx context.Context, w http.ResponseWriter, sctx *Sessi
 	return nil
 }
 
+func clearSessionCookie(w http.ResponseWriter) {
```
nit: Plural.

```suggestion
func clearSessionCookies(w http.ResponseWriter) {
```

codingllama replied:
godoc please, it's worth mentioning which cookies it clears.

codingllama commented on lib/web/apiserver.go:
```diff
@@ -4061,9 +4080,10 @@ func (h *Handler) AuthenticateRequest(w http.ResponseWriter, r *http.Request, ch
 	if err != nil {
 		return nil, trace.AccessDenied("failed to decode cookie")
 	}
+	// sshah
```
Leftover comment?

codingllama commented on lib/idp/saml/session.go:
```diff
@@ -0,0 +1,52 @@
+	SAMLSessionCookieName = "__Host-saml_session"
+)
+
+// SetCookie set's SAML session cookie named __Host-saml_session.
```
Suggestion:

```suggestion
// SetCookie set's the cookie named by [SAMLSessionCookieName].
```

codingllama commented on lib/idp/saml/session.go:
```diff
@@ -0,0 +1,52 @@
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package saml
```
Thanks, the separation makes both files/packages clearer and more concise.

codingllama commented on lib/web/apiserver_test.go:
```diff
@@ -8965,6 +8966,74 @@ func TestLogout(t *testing.T) {
+}
+
+func hasEmptySAMLSessionCookieValue(cookies []*http.Cookie) bool {
+	samlCookie := &http.Cookie{
+		Name:     samlidp.SAMLSessionCookieName,
+		Value:    "",
+		Path:     "/",
+		HttpOnly: true,
+		Secure:   true,
+	}
+	for _, cookie := range cookies {
+		if cookie.String() == samlCookie.String() {
```
Maybe:

```suggestion
	samlCookie := (&http.Cookie{
		Name:     samlidp.SAMLSessionCookieName,
		Value:    "",
		Path:     "/",
		HttpOnly: true,
		Secure:   true,
	}).String()
	for _, cookie := range cookies {
		if cookie.String() == samlCookie {
```

So we only calculate this string once?

codingllama commented on lib/web/sessions.go:
```diff
@@ -850,6 +850,19 @@ func (s *sessionCache) invalidateSession(ctx context.Context, sctx *SessionConte
 	}
+
+	if err := clt.DeleteUserAppSessions(ctx, &proto.DeleteUserAppSessionsRequest{Username: sctx.GetUser()}); err != nil {
+		return trace.Wrap(err)
```
I realize this is pre-existing, but I wonder how correct it is to abort deletion halfway. Maybe we should do all deletions once this starts, and then return the joined error from all outcomes.

WDYT @flyinghermit and reviewers?

mdwn replied:
IMO that seems like a good move. Otherwise we won't best-effort try to get rid of everything.

flyinghermit replied:
Yeah I was just being lazy hoping someone else would take that effort. Updated it so the error is combined in https://github.com/gravitational/teleport/pull/37619/commits/8efeb62deca926edc0a9118b926c5c466ba50919
lmk if that's the correct way to do that!

codingllama replied:
Thanks for taking the plunge. I would use an `[]error` for the impl (see comments) but that's kind of minor.

codingllama commented on lib/web/sessions.go:
```diff
@@ -850,6 +850,19 @@ func (s *sessionCache) invalidateSession(ctx context.Context, sctx *SessionConte
+		return trace.Wrap(err)
+	}
+
+	if sctx.cfg.Session.GetSAMLSession() != nil {
+		if err := clt.DeleteSAMLIdPSession(ctx, types.DeleteSAMLIdPSessionRequest{
+			SessionID: sctx.cfg.Session.GetSAMLSession().ID,
+		}); err != nil && !trace.IsNotFound(err) {
+			return trace.Wrap(err)
+		}
+	}
```
Suggestion: Use the if scope to cut repetition, check the ID as well.

```suggestion
	if session := sctx.cfg.Session.GetSAMLSession(); session != nil && session.ID != "" {
		if err := clt.DeleteSAMLIdPSession(ctx, types.DeleteSAMLIdPSessionRequest{
			SessionID: session.ID,
		}); err != nil && !trace.IsNotFound(err) {
			return trace.Wrap(err)
		}
	}
```

codingllama commented:
@mdwn, as the expert I think you should take a look at this one. I can't say I know the inner workings of the SAML IdP too well.

mdwn reviewed (commented):
No issue with the approach here. Overall this LGTM, but want to make sure the rest of the feedback is addressed.

flyinghermit committed:
address review comments

flyinghermit committed:
combine app, saml and web session deletion error

codingllama commented on lib/web/sessions.go:
```diff
@@ -850,18 +851,34 @@ func (s *sessionCache) invalidateSession(ctx context.Context, sctx *SessionConte
+	// App session, SAML session and web session deletion should be treated as a single transaction.
+	// To avoid aborting deletion midpoint due to a failure in one of the session deletion,
+	// we use sessionDeletionErr below to join errors and return them at last.
+	var sessionDeletionErr error
```
Suggestion:

```suggestion
	var deleteErrs []error
```

and then append to it? This gives us a single, "flat" joined error instead of a joined tree. We can also rely on errors.Join to decide if the result is `nil`.

See the following comments for examples.

flyinghermit replied:
@codingllama Curios, what would be pros of "flat" error over joined ones? 

> errors.Join to decide if the result is nil

Yeah but we also check for if err is `NotFound` so we aren't doing much.

codingllama replied:
The snippet would go something like this:

```go

	// App session, SAML session and web session deletion should be treated as a single transaction.
	// To avoid aborting deletion midpoint due to a failure in one of the session deletion,
	// we use sessionDeletionErr below to join errors and return them at last.
	var deleteErrs error
	if err := clt.DeleteUserAppSessions(ctx, &proto.DeleteUserAppSessionsRequest{Username: sctx.GetUser()}); err != nil {
		deleteErrs = append(deleteErrs, err)
	}
	if samlSession := sctx.cfg.Session.GetSAMLSession(); samlSession != nil && samlSession.ID != "" {
		if err := clt.DeleteSAMLIdPSession(ctx, types.DeleteSAMLIdPSessionRequest{
			SessionID: samlSession.ID,
		}); err != nil && !trace.IsNotFound(err) {
			deleteErrs = append(deleteErrs, err)
		}
	}
	// Delete just the session - leave the bearer token to linger to avoid
	// failing a client query still using the old token.
	if err := clt.WebSessions().Delete(ctx, types.DeleteWebSessionRequest{
		User:      sctx.GetUser(),
		SessionID: sctx.GetSessionID(),
	}); err != nil && !trace.IsNotFound(err) {
		deleteErrs = append(deleteErrs, err)
	}

	if err := errors.Join(deleteErrs...); err != nil {
		return trace.Wrap(deleteErrs)
	}
```

We could also drop most ifs and always join, like this:

```go

	// App session, SAML session and web session deletion should be treated as a single transaction.
	// To avoid aborting deletion midpoint due to a failure in one of the session deletion,
	// we use sessionDeletionErr below to join errors and return them at last.
	var deleteErrs error
	err = clt.DeleteUserAppSessions(ctx, &proto.DeleteUserAppSessionsRequest{Username: sctx.GetUser()})
	deleteErrs = append(deleteErrs, err)

	if samlSession := sctx.cfg.Session.GetSAMLSession(); samlSession != nil && samlSession.ID != "" {
		if err := clt.DeleteSAMLIdPSession(ctx, types.DeleteSAMLIdPSessionRequest{
			SessionID: samlSession.ID,
		}); err != nil && !trace.IsNotFound(err) {
			deleteErrs = append(deleteErrs, err)
		}
	}

	// Delete just the session - leave the bearer token to linger to avoid
	// failing a client query still using the old token.
	err = clt.WebSessions().Delete(ctx, types.DeleteWebSessionRequest{
		User:      sctx.GetUser(),
		SessionID: sctx.GetSessionID(),
	})
	deleteErrs = append(deleteErrs, err)

	if err := errors.Join(deleteErrs...); err != nil {
		return trace.Wrap(deleteErrs)
	}
```

flyinghermit replied:
> A "flat" join would traverse with less hops. It's not a big deal.

Thanks for explaining and reference code. In the second example, we can only skip `if` block in the `DeleteUserAppSessions`.  For the first example, Since we are calling three different delete func, as opposed to calling one delete func thrice, I don't think we are gaining much here.

From the `errors.Join()` godoc:
> // The error formats as the concatenation of the strings obtained
> // by calling the Error method of each element of errs, with a newline
> // between each string.

codingllama commented on lib/web/sessions.go:
```diff
@@ -850,18 +851,34 @@ func (s *sessionCache) invalidateSession(ctx context.Context, sctx *SessionConte
+	// we use sessionDeletionErr below to join errors and return them at last.
+	var sessionDeletionErr error
+	if err := clt.DeleteUserAppSessions(ctx, &proto.DeleteUserAppSessionsRequest{Username: sctx.GetUser()}); err != nil {
+		sessionDeletionErr = err
```
```suggestion
		deleteErrors = append(deleteErrs, err)
```

codingllama commented on lib/web/sessions.go:
```diff
@@ -850,18 +851,34 @@ func (s *sessionCache) invalidateSession(ctx context.Context, sctx *SessionConte
 	}
-	if err := clt.DeleteUserAppSessions(ctx, &proto.DeleteUserAppSessionsRequest{Username: sctx.GetUser()}); err != nil {
-		return trace.Wrap(err)
+
+	if sessionDeletionErr != nil {
+		return trace.Wrap(sessionDeletionErr)
```
```suggestion
	if err := errors.Join(deleteErrs...); err != nil {
		return trace.Wrap(err)
```

this way we can let errors.Join figure it out for us.

flyinghermit replied:
This looks counter intuitive to me. We are just doing two joins, append to array and then join those error again, instead of joining error once.

flyinghermit committed:
remove debug logging in case saml cookie is not found

flyinghermit commented:
Friendly ping @Tener @mdwn

flyinghermit committed:
update TODO comment for updating SAML session

flyinghermit committed:
fix comment typo udpate -> update
